File: web/src/synthesis/voicesBin.ts
import { PolyReadError } from "../core/errors";
import { Voice, VOICE_DIMENSION, VOICE_ROWS } from "./voices";

/**
 * §7.1's packed format, kept for anyone who already has the iOS build's
 * `Voices.bin`: "24-byte header (magic `"AIOSVOX"` + version byte, count,
 * lengths=510, dimension=256, reserved), then `24·n` NUL-padded ASCII names,
 * then `count · 510 · 256` float32 LE."
 *
 * The web distribution uses one bare file per voice instead (see `voices.ts`),
 * so nothing in the app reaches for this unless a packed file is configured.
 */
const MAGIC = "AIOSVOX";
const HEADER_SIZE = 24;
const NAME_SIZE = 24;

export interface VoicesBinHeader {
  version: number;
  count: number;
  lengths: number;
  dimension: number;
}

export class VoicesBin {
  readonly header: VoicesBinHeader;
  readonly names: string[];
  private readonly styles: Float32Array;

  constructor(data: ArrayBuffer) {
    if (data.byteLength < HEADER_SIZE) {
      throw new PolyReadError(
        "voicesFileMalformed",
        `file is ${data.byteLength} bytes, shorter than the header`,
      );
    }
    const bytes = new Uint8Array(data);
    const view = new DataView(data);

    const magic = new TextDecoder().decode(bytes.subarray(0, 7));
    if (magic !== MAGIC) {
      throw new PolyReadError("voicesFileMalformed", `magic is "${magic}", expected "${MAGIC}"`);
    }

    const version = bytes[7];
    const count = view.getUint32(8, true);
    const lengths = view.getUint32(12, true);
    const dimension = view.getUint32(16, true);

    if (count <= 0 || lengths <= 0 || dimension <= 0) {
      throw new PolyReadError(
        "voicesFileMalformed",
        `count=${count} lengths=${lengths} dimension=${dimension}`,
      );
    }

    const namesSize = count * NAME_SIZE;
    const floatCount = count * lengths * dimension;
    const expected = HEADER_SIZE + namesSize + floatCount * 4;
    if (data.byteLength < expected) {
      throw new PolyReadError(
        "voicesFileMalformed",
        `file is ${data.byteLength} bytes, expected at least ${expected} for ${count} voices`,
      );
    }

    const names: string[] = [];
    const decoder = new TextDecoder();
    for (let i = 0; i < count; i++) {
      const start = HEADER_SIZE + i * NAME_SIZE;
      const slice = bytes.subarray(start, start + NAME_SIZE);
      let end = slice.indexOf(0);
      if (end < 0) end = slice.length;
      names.push(decoder.decode(slice.subarray(0, end)));
    }

    const floatStart = HEADER_SIZE + namesSize;
    // The file is little-endian and every platform this runs on is too, so the
    // view is already correct; there is nothing to byte-swap.
    this.styles = new Float32Array(data.slice(floatStart, floatStart + floatCount * 4));
    this.header = { version, count, lengths, dimension };
    this.names = names;
  }

  voice(name: string): Voice {
    const index = this.names.indexOf(name);
    if (index < 0) throw new PolyReadError("voiceNotFound", name);
    const { lengths, dimension } = this.header;
    if (lengths !== VOICE_ROWS || dimension !== VOICE_DIMENSION) {
      throw new PolyReadError(
        "voicesFileMalformed",
        `lengths=${lengths} dimension=${dimension}, expected ${VOICE_ROWS}x${VOICE_DIMENSION}`,
      );
    }
    const start = index * lengths * dimension;
    return new Voice(name, this.styles.slice(start, start + lengths * dimension));
  }
}

File: Sources/PolyReadSynthesis/VoicesBin.swift
import Foundation
import PolyReadCore

/// §7.1 — "`Voices.bin`: 24-byte header (magic `"AIOSVOX"` + version byte, count,
/// lengths=510, dimension=256, reserved), then `24·n` NUL-padded ASCII names,
/// then `count · 510 · 256` float32 LE. **Style row index = phoneme count
/// excluding the two boundary tokens.**"
public struct VoicesBin: Sendable {

    public struct Header: Sendable, Equatable {
        public let version: UInt8
        public let count: Int
        public let lengths: Int      // 510
        public let dimension: Int    // 256
    }

    public let header: Header
    public let names: [String]
    /// Flat: `voice * lengths * dimension + row * dimension + component`.
    private let styles: [Float]

    static let magic = "AIOSVOX"
    static let headerSize = 24
    static let nameSize = 24

    public init(data: Data) throws {
        guard data.count >= Self.headerSize else {
            throw PolyReadError.voicesFileMalformed("file is \(data.count) bytes, shorter than the header")
        }

        let magic = String(decoding: data[0..<7], as: UTF8.self)
        guard magic == Self.magic else {
            throw PolyReadError.voicesFileMalformed("magic is \"\(magic)\", expected \"\(Self.magic)\"")
        }

        let version = data[data.startIndex + 7]
        let count = Int(Self.readUInt32(data, at: 8))
        let lengths = Int(Self.readUInt32(data, at: 12))
        let dimension = Int(Self.readUInt32(data, at: 16))

        guard count > 0, lengths > 0, dimension > 0 else {
            throw PolyReadError.voicesFileMalformed("count=\(count) lengths=\(lengths) dimension=\(dimension)")
        }

        let namesSize = count * Self.nameSize
        let floatCount = count * lengths * dimension
        let expected = Self.headerSize + namesSize + floatCount * MemoryLayout<Float>.size
        guard data.count >= expected else {
            throw PolyReadError.voicesFileMalformed(
                "file is \(data.count) bytes, expected at least \(expected) for \(count) voices"
            )
        }

        var names: [String] = []
        names.reserveCapacity(count)
        for i in 0..<count {
            let start = data.startIndex + Self.headerSize + i * Self.nameSize
            let bytes = data[start..<(start + Self.nameSize)]
            let trimmed = bytes.prefix { $0 != 0 }
            names.append(String(decoding: trimmed, as: UTF8.self))
        }

        let floatStart = data.startIndex + Self.headerSize + namesSize
        var styles = [Float](repeating: 0, count: floatCount)
        styles.withUnsafeMutableBytes { destination in
            data.copyBytes(
                to: destination.bindMemory(to: UInt8.self),
                from: floatStart..<(floatStart + floatCount * MemoryLayout<Float>.size)
            )
        }
        // The file is little-endian and every device this runs on is too, so the
        // copy above is already correct; there is nothing to byte-swap.

        self.header = Header(version: version, count: count, lengths: lengths, dimension: dimension)
        self.names = names
        self.styles = styles
    }

    public init(url: URL) throws {
        try self.init(data: try Data(contentsOf: url, options: .mappedIfSafe))
    }

    public static func bundled(bundle: Bundle = .main) throws -> VoicesBin {
        guard let url = bundle.url(forResource: "Voices", withExtension: "bin") else {
            throw PolyReadError.modelMissing("Voices.bin")
        }
        return try VoicesBin(url: url)
    }

    /// The style vector for a voice at a given phoneme count.
    ///
    /// §7.1's rule is that the row index *is* the phoneme count with the two
    /// boundary zeros excluded — which is also why §6.2 wants uniform chunk
    /// lengths: wildly varying counts pick wildly varying style rows and the
    /// prosody wanders.
    public func style(voice name: String, phonemeCount: Int) throws -> [Float] {
        guard let index = names.firstIndex(of: name) else {
            throw PolyReadError.voiceNotFound(name)
        }
        // A chunk exactly at the 510 budget would index one past the last row.
        let row = min(max(0, phonemeCount), header.lengths - 1)
        let start = (index * header.lengths + row) * header.dimension
        return Array(styles[start..<(start + header.dimension)])
    }

    private static func readUInt32(_ data: Data, at offset: Int) -> UInt32 {
        let start = data.startIndex + offset
        var value: UInt32 = 0
        for i in 0..<4 {
            value |= UInt32(data[start + i]) << (8 * UInt32(i))
        }
        return value
    }
}
